Notification tray for displaying in-flight requests

somebody1234 opened the pull request:
### Pull Request Description
- Close https://github.com/enso-org/cloud-v2/issues/1641
  - Implement notification tray to display all in-flight mutations

### Important Notes
- This PR uses `useIsMutating` to find mutation state. It's possible to use `useIsFetching` to also display query state - however "list directory" background refetching would cause constant notifications.
- UI is a placeholder - it does not need to be a notification tray, the important part of this PR is the new hook to return the state of all in-flight mutations.

### Testing Instructions
- Do all mutation types (delete, rename, copy, move, undelete, force delete, clear trash)
- Open notification tray
- Check for notification
  - The notification may have disappeared if the mutation has completed - this is intentional as the notifications are not meant to be permanent.

### Further Improvements
- Display red "unread notifications" badge when new notifications are in flight
- Consider creating permanent notification objects instead, which include a timestamp, and are (possibly) preserved indefinitely
  - If the user is (e.g.) uploading or deleting a lot of assets though, it would clutter up the notification tray if/unless notifications caused by mutations are collapsed.
  - Notifications do not exist server side, so they are not preserved across sessions. This is fine, but we might want to have a log of mutations on the server side eventually.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

somebody1234 committed:
WIP

somebody1234 committed:
WIP: Initial implementation of transient notifications and notification tray

somebody1234 committed:
WIP: Fix bugs

somebody1234 committed:
WIP

somebody1234 committed:
WIP: Debugging

somebody1234 committed:
Remove unused dependency

somebody1234 committed:
Add support for colors in `NotificationItem`

somebody1234 committed:
help

somebody1234 committed:
WIP: Fix RAC error

somebody1234 committed:
Fix issue with `GridList` crashing entire app

somebody1234 committed:
Remove testing notification

somebody1234 committed:
Remove inaccurate notification count as mutations are batched

somebody1234 committed:
Implement notification for file uploads progress

somebody1234 committed:
Fix Bazel on NixOS

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Fix Bazel build

somebody1234 committed:
Initial UI for file upload notification

somebody1234 committed:
WIP: Make computed notifications permanent

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Make notifications persistent

somebody1234 committed:
Fix user bar styling

somebody1234 committed:
Omit special assets from being selected

somebody1234 committed:
Fix notification display

somebody1234 committed:
Add timestamp to notifications

somebody1234 committed:
Show timestamp in notification entry

somebody1234 committed:
Add button to Devtools to clear TanStack query cache and reload

somebody1234 committed:
Show red indicator on unread notifications; fix notification display; show notifications by date descending

somebody1234 committed:
pnpm i

somebody1234 committed:
Fix errors

PabloBuchu commented:
1. ~~Styling issues? The `+` changed color~~
<img width="136" alt="Screenshot 2025-02-05 at 10 24 24" src="https://github.com/user-attachments/assets/9c1f12ce-ce94-4cb9-a966-d4a33b3c83bc" />

2. Upload image is completely broken. This is what I am getting when I drop a file
<img width="1719" alt="Screenshot 2025-02-05 at 10 25 02" src="https://github.com/user-attachments/assets/668f690b-93d4-490a-928d-7b1eeefc7e02" />

3.  The notification try. Like that we can use that later for backend related stuff as well. I think maybe we should move it to the user context menu and display small number on the avatar indicating some unread messages? How can I trigger a message to appear in the tray?
<img width="99" alt="Screenshot 2025-02-05 at 10 26 49" src="https://github.com/user-attachments/assets/018e8064-f6c2-459a-ae16-6f1746f9e89a" />

4. Create >10 directories -> select them all -> hit delete -> accept dialog. The loader cursor appears for a fraction of time but when I am waiting for dirs to be removed I loader is not here (they are disabled yes) but it feels inconsistent.

somebody1234 committed:
Fix infinite loop when uploading file

somebody1234 committed:
Support indeterminate progress state for notifications

somebody1234 committed:
Disable dropzone in Trash and Recent categories

somebody1234 commented:
1. :information_source: out of scope
2. :white_check_mark: Upload image is completely broken
    - fixed and pushed
4. .
   - :information_source: we can use that later for backend related stuff as well
     - yes, the idea is that it's just an array of objects that are shown by components. we should be able to nearly trivially query for a separate array from the backend and then merge the arrays together. ideally they should have timestamps so they can be sorted in the correct order
   - :warning: maybe we should move it to the user context menu and display small number on the avatar indicating some unread messages?
     - sounds reasonable, will look into it soon:tm:
   - :information_source: How can I trigger a message to appear in the tray?
     - right now? no backend support, but batch copy, batch move, batch delete, batch undelete mutations are tracked
       - single copy/move/delete/undeletes also show up! because we recently made them use the batched methods for simplicity's sake
5. :white_check_mark: Create >10 directories -> select them all -> hit delete -> accept dialog. The loader cursor appears for a fraction of time but when I am waiting for dirs to be removed I loader is not here
    - fixed and pushed

somebody1234 committed:
WIP: Fix "confirm delete" modal

somebody1234 committed:
Actually stop "confirm delete" modal from closing early when deleting assets...

somebody1234 committed:
Add close button to notifications

somebody1234 committed:
Switch `NotificationItem` to use `tailwind-variants`

somebody1234 committed:
Use icon instead of button for notification icon

somebody1234 committed:
WIP: Show toast for some notifications

somebody1234 committed:
Fix crash

somebody1234 committed:
Fix notification toasts

somebody1234 committed:
Remove `useUploadFileWithToastMutation` because notifications now update toasts

somebody1234 committed:
Distinguish hard delete and soft delete in notifications

somebody1234 committed:
Fix detection for whether a notification is finished

somebody1234 committed:
Fix notification toast display and autoclose

somebody1234 committed:
Fix layout of close button for notifications

somebody1234 committed:
`ProgressBar` component

somebody1234 committed:
UI for indeterminate progress bar state

somebody1234 commented:
@MrFlashAccount what was consensus on the solution to not wanting to display all of these temporary notifications btw?
- delete after 5 min?
- only show most recent?
- keep for now and wait for feedback?
- or maybe implement one of the two and add alternate modes behind feature flags?

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

MrFlashAccount commented on app/gui/src/dashboard/components/AriaComponents/ProgressBar/ProgressBar.tsx:
```diff
@@ -0,0 +1,44 @@
+}
+
+/** Progress bar. */
+export function ProgressBar(props: ProgressBarProps) {
```
Please add stories

somebody1234 replied:
added, lmk if there are more cases i should add and/or adjustments i should make to the component

MrFlashAccount commented on app/gui/src/dashboard/components/AriaComponents/Button/variants.ts:
```diff
@@ -28,6 +28,16 @@ export const BUTTON_STYLES = tv({
     'transition-[opacity,outline-offset,background,border-color] duration-150 ease-in-out',
   ],
   variants: {
+    color: {
```
What's the purpose of this?

MrFlashAccount commented on app/gui/src/dashboard/assets/inbox_with_notification.svg:
```diff
@@ -0,0 +1,5 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
+<path d="M5 9H2V13C2 13.5523 2.44772 14 3 14H13C13.5523 14 14 13.5523 14 13V9H11C10.4477 9 10 9.44772 10 10C10 11.1046 9.10457 12 8 12C6.89543 12 6 11.1046 6 10C6 9.44772 5.55228 9 5 9Z" fill="black" stroke="black"/>
+<path fill-rule="evenodd" clip-rule="evenodd" d="M10.5351 0C10.1948 0.588347 10 1.27143 10 2H3C2.44772 2 2 2.44772 2 3V13C2 13.5523 2.44772 14 3 14H13C13.5523 14 14 13.5523 14 13V6C14.7286 6 15.4117 5.80521 16 5.46487V13C16 14.6569 14.6569 16 13 16H3C1.34315 16 0 14.6569 0 13V3C0 1.34315 1.34315 0 3 0H10.5351Z" fill="black"/>
+<circle cx="14" cy="2" r="2" fill="#F3180A"/>
```
The red dot should be a component, similarly to: https://chakra-ui.com/docs/components/avatar#badge

somebody1234 replied:
added as `StatusBadge`

MrFlashAccount commented on app/gui/src/dashboard/assets/inbox.svg:
```diff
@@ -0,0 +1,4 @@
+<svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Where did you pick this icon up? Not sure it's coming from the central icon system

somebody1234 replied:
made it manually... not ideal but i'm just really not sure it fits in with that amount of padding
![image](https://github.com/user-attachments/assets/6a8f37b0-a600-4723-8885-a5c19a222bfb)

somebody1234 replied:
the padding also means badges might look weird

MrFlashAccount commented on app/common/src/text.ts:
```diff
@@ -143,12 +137,32 @@ interface PlaceholderOverrides {
   readonly xItemsCopied: [count: number]
   readonly xItemsCut: [count: number]
-  readonly ordinalFallback: [number: number]
+  readonly deletingXAssetsNotification: [count: number]
```
I think we can remove it

MrFlashAccount replied:
@somebody1234 👆

MrFlashAccount commented:
@somebody1234 the notifications tray only displays _ongoing_ actions, list cleans up automatically when action finishes

somebody1234 committed:
Only show notifications for uploads by default

somebody1234 committed:
Use central icon for inbox

somebody1234 committed:
Remove `color` prop from `Button` component

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Clean up finished computed notification 1 minute after action finishes

somebody1234 committed:
Add and use `StatusBadge` component

somebody1234 committed:
Add stories for `ProgressBar`

somebody1234 committed:
Add stories for `StatusBadge`

somebody1234 commented:
@MrFlashAccount ready for re-CR

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[ts] Fix type errors

somebody1234 committed:
[ts] Fix lint errors

somebody1234 committed:
[dashboard] Fix integration tests?

MrFlashAccount commented:
QA/UX:

1. When upload error happens, the notification with upload doesn't show the proper status. (still in loading state)
2. I can't remove the items from the list if an error happens.
3. During upload close button should be hidden, show cancel button below instead.
4. gaps between the inbox button and other elements should be larger - 6px.
5. Upload notfication should appear below the inbox button. Disappear after 5secs.
6. Maybe we should hide the inbox when there's no items?
7. We need to indicate error state either. When error happens there should be close(cross) and retry buttons.
8.  If user goes offline, we need to indicate it either.
9. I think notification tray should be the most left button.
10. Upload to local backend should show indeterminate (https://react-spectrum.adobe.com/react-aria/ProgressBar.html#indeterminate) progress bar.
12. Do we display the downloads there?
13. prefer different icons for different states(empty - outline, loading - icon filled inside a loaded. error - filled with a red circle around.
14. Let's keep the items in the list until user discard them manually (sorry for the prev message: https://github.com/enso-org/enso/pull/12078#issuecomment-2647483185)
15. Each file in upload should be a separate notification item

PS Is it still notifications? 

<blockquote><img src="https://react-spectrum.adobe.com/ReactAriaOpenGraph.6ea0ee74.webp" width="48" align="right"><div><strong><a href="https://react-spectrum.adobe.com/react-aria/ProgressBar.html">ProgressBar</a></strong></div><div>Documentation for ProgressBar in the React Aria package.</div></blockquote>

MrFlashAccount commented on app/gui/src/dashboard/components/AriaComponents/ProgressBar/ProgressBar.tsx:
```diff
@@ -0,0 +1,44 @@
+})
+
+/** Props for a {@link ProgressBar}. */
+export interface ProgressBarProps extends Omit<AriaProgressBarProps, 'value'> {
```
extend `VariantProps<typeof PROGRESS_BAR_STYLES>`

MrFlashAccount commented on app/gui/src/dashboard/components/AriaComponents/ProgressBar/ProgressBar.tsx:
```diff
@@ -0,0 +1,44 @@
+  const { progress, ...rest } = props
+  const progressNumber = progress === 'indeterminate' ? 1 : progress
+
+  const styles = PROGRESS_BAR_STYLES()
```
Don't forget that `VariantProps<typeof PROGRESS_BAR_STYLES>` allows customizing `variants`

MrFlashAccount commented on app/gui/src/dashboard/components/AriaComponents/StatusBadge/StatusBadge.tsx:
```diff
@@ -0,0 +1,45 @@
+export function StatusBadge(props: StatusBadgeProps) {
+  const { color, hidden, children } = props
+
+  const styles = STATUS_BADGE_STYLES({ color, hidden })
```
Same as previous

MrFlashAccount commented on app/gui/src/dashboard/layouts/AssetsTable.tsx:
```diff
@@ -288,6 +288,21 @@ function AssetsTable(props: AssetsTableProps) {
   const lastSelectedIdsRef = useRef<AssetId | ReadonlySet<AssetId> | null>(null)
   const headerRowRef = useRef<HTMLTableRowElement>(null)
   const getPasteData = useEventCallback(() => driveStore.getState().pasteData)
+  const canUploadToThisCategory = (() => {
```
I think this should be extracted into a separate function

MrFlashAccount replied:
I like when things are `prebuilt` - for example, I put `icon` for each category into `category` object. Maybe we can do the same for this flag?

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/components/NotificationItem.tsx:
```diff
@@ -0,0 +1,52 @@
+}
+
+/** An item in the notification tray. */
+export function NotificationItem(props: NotificationItemProps) {
```
FWIW, we can use storybook for this component aswell

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+/** Return the number of ongoing mutations of the given type across both backends. */
+export function useIsMutatingForBothBackends(makeKey: (backendType: BackendType) => MutationKey) {
+  return (
+    useIsMutating({ mutationKey: makeKey(BackendType.local) }) +
```
Here you can use `predicate`: https://tanstack.com/query/latest/docs/framework/react/guides/filters#mutation-filters and get rid of the second `useIsMutating`

somebody1234 replied:
idk whether i'm a fan of this. i've thought about it before, but according to the linked docs:
> If no other filters are specified, this function will be evaluated against every mutation in the cache.

so i'd take my chances on matching by key being somewhat optimized (at least eventually) vs `predicate` which is impossible to optimize

MrFlashAccount replied:
hmm, let's keep it as-is then

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+  )
+
+  const removeComputedNotification = useEventCallback((id: string) => {
+    setNotificationMap((map) => new Map([...map.entries()].filter(([, v]) => v.id !== id)))
```
```
map.delete(id)
return new Map(map)
```

somebody1234 replied:
mutation, but yeah the opposite order would work

somebody1234 replied:
wait no, `.delete(id)` does not work because `id` is not the key

MrFlashAccount replied:
why can't we use it as a key though?

somebody1234 replied:
currently the mutation variables (a stable reference to an object) are used as the key.

that said, i did add a function to convert the stable reference to a string id. thoughts? should we be keying the map on the string id instead then?

MrFlashAccount replied:
It depends, do we more often query by id or by the mutation variable reference?

somebody1234 replied:
once we have the notification object itself then we can easily query by id, the mutation variables are only really visible internally. so i have no problem switching over to id

somebody1234 replied:
although it's worth noting that converting to id would happen internally anyway so to the consumer it's invisible either way. and the object -> string mappings are kept in a map anyway so it's not particularly expensive no matter where the lookup is done

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+                  message:
+                    force ?
+                      getText(
+                        isSuccess ? 'permanentlyDeletedXAssetsNotification'
```
no nested ternaries please

MrFlashAccount replied:
It'd awesome to turn on the https://eslint.org/docs/latest/rules/no-nested-ternary rule

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+          existingNotification?.timestamp ?? newNotification.timestamp ?? Number(new Date()),
+      }
+      newNotifications.set(key, notification)
+      const isFinished =
```
looks dirty, maybe there's a way to simplify it?

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+        })
+      }
+      if (isFinished) {
+        setTimeout(() => {
```
settimeout without cleanup is a bad sign

somebody1234 replied:
oh right i haven't addressed this yet (but this should be the final thing that needs to be addressed from this round). not sure what alternative solutions there are

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/computedNotificationHooks.tsx:
```diff
@@ -0,0 +1,287 @@
+  })
+
+  useEffect(() => {
+    if (moreComputedNotifications) {
```
I think it should be extracted from this hook, as it's still under consideration.

MrFlashAccount commented on app/gui/src/dashboard/layouts/NotificationTray/types.ts:
```diff
@@ -0,0 +1,15 @@
+/** @file Types related to the `NotificationTray`. */
```
Can we move it under `pages/dashboard/components` folder?

MrFlashAccount commented on app/gui/src/dashboard/pages/dashboard/DashboardTabBar.tsx:
```diff
@@ -43,47 +45,56 @@ export function DashboardTabBar(props: DashboardTabBarProps) {
 
   return (
-    <TabBar className="bg-primary/10" items={tabs}>
+    <TabBar className="bg-primary/5" items={tabs}>
```
we changed it to /5, why? By design it's /10

somebody1234 replied:
dunno :sweat_smile: will fix

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard/DashboardTabBar] Fix background opacity

somebody1234 committed:
[dashboard/NotificationTray] Move to new folder

somebody1234 committed:
[dashboard/computedNotificationHooks] Address CR

somebody1234 committed:
[dashboard] Use `variants` prop instead of `_STYLES` constants directly

somebody1234 committed:
[dashboard/ProgressBar] Improve storybook

somebody1234 committed:
[dashboard/NotificationItem] Add storybook

somebody1234 committed:
[dashboard/NotificationItem] Improve storybook

somebody1234 committed:
[dashboard/Icon] Improve storybook

somebody1234 committed:
[dashboard/Breadcrumbs] Rename storybook stories

somebody1234 committed:
[dashboard/Text] Improve storybook

somebody1234 committed:
[dashboard/Text] Add `StoryVariants` component

somebody1234 committed:
[dashboard/ProgressBar] Use `StoryVariants`

somebody1234 committed:
[dashboard/Icon] Use `StoryVariants`

somebody1234 committed:
[dashboard/NotificationItem] Improve type safety in stories

somebody1234 committed:
[dashboard/Button] Fix type error in stories

somebody1234 committed:
[dashboard/Check] Use `StoryVariants`

somebody1234 committed:
[dashboard/Button] Use `StoryVariants`

somebody1234 committed:
[dashboard/StatusBadge] Use `StoryVariants`

somebody1234 committed:
[dashboard/ComboBox] Use `StoryVariants`

somebody1234 committed:
[dashboard/DatePicker] Use `StoryVariants`

somebody1234 committed:
[dashboard/Selector] Use `StoryVariants`

somebody1234 committed:
[dashboard/computedNotificationHooks] Fix type error

somebody1234 committed:
[dashboard/Category] Add `canUploadHere` field for drag and drop

somebody1234 committed:
[dashboard/NotificationItem] Fix type error in stories

somebody1234 committed:
[dashboard/Button] Prettier

somebody1234 committed:
[dashboard/LocalStorage] Fix type errors

somebody1234 committed:
[dashboard/LocalStorage] Fix type errors

somebody1234 committed:
[dashboard/LocalStorage] will this make lints pass?

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard] Remove unnecessary extra computed notifications

MrFlashAccount commented:
@somebody1234 is it ready for CR/QA?

somebody1234 commented:
@MrFlashAccount i hope so! there's a lot of CR items so not all of it might be addressed, but I think it might be better to get a re-CR bc i *think* most should be addressed and i'm not 100% which specifically

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard] Fix type errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard] Fix lint errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard] Prettier

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard/projectHooks] Fix invalid import

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
Merge branch 'develop' into wip/sb/loader-for-requests

somebody1234 committed:
[dashboard/ModifiedColumn] Avoid line wrap
